src/track/beats.cpp

```diff
@@ -663,30 +861,260 @@ std::optional<BeatsPointer> Beats::tryScale(BpmScale scale) const {
+    double adjustmentFactor = static_cast<int>(adjustment) & 0b100 ? kBpmLargeAdjustModifier : 1;
+    adjustment = static_cast<TempoAdjustment>(static_cast<int>(adjustment) & 0b11);
+    if (markerIt == markers.end()) {
+        lastMarkerBpm +=
+                ((adjustment == TempoAdjustment::Faster) ? kBpmAdjustStep
+                                                         : -kBpmAdjustStep) *
+                adjustmentFactor;
+        if (!lastMarkerBpm.isValid()) {
+            qWarning() << "Beats: Tempo adjustment would result in invalid Bpm!";
+            return std::nullopt;
+        }
+    } else {
+        const auto marker = *markerIt;
+        double currentMarkerBpm = (60.0 * m_sampleRate) / marker.beatsLength();
+        currentMarkerBpm += ((adjustment == TempoAdjustment::Faster)
+                                            ? kBpmAdjustStep
+                                            : -kBpmAdjustStep) *
+                adjustmentFactor;
+        const double adjustedBeatLength = (60.0 * m_sampleRate) / (currentMarkerBpm);
+
+        if (adjustedBeatLength <= 0) {
+            qWarning() << "Beats: Tempo adjustment would result in a marker "
+                          "with less than the minimum beats in one bar!";
+            return std::nullopt;
+        }
+
+        markerIt = markers.erase(markerIt);
+        markerIt = markers.emplace(markerIt,
+                marker.position(),
+                adjustedBeatLength,
+                marker.beatsPerBar());
+    }
+
+    return fromBeatMarkers(m_sampleRate,
+            std::move(markers),
+            m_lastMarkerPosition,
+            m_lastBeatsPerBar,
+            lastMarkerBpm);
+}
+
+std::optional<BeatsPointer> Beats::trySetMarker(audio::FramePos position) const {
+    auto markers = m_markers;
+    auto markerIt = std::lower_bound(markers.begin(),
+            markers.end(),
+            position,
+            isBeatMarkerLessThanFramePos);
+
+    DEBUG_ASSERT(!m_lastMarkerPosition.isFractional());
+    auto lastMarkerPosition = m_lastMarkerPosition;
+    auto lastMarkerBeatsPerBar = m_lastBeatsPerBar;
+    if (markerIt == markers.end()) {
+        const int numBeats = calculateBeatsTillNextMarker(
+                iteratorClosestTo(position), lastMarkerPosition);
+        if (numBeats == 0) {
+            // The new marker would be too close to the existing marker, so we
+            // just move the existing marker instead.
+            lastMarkerPosition = position.toLowerFrameBoundary();
+        } else if (lastMarkerPosition < position) {
+            // We are behind the last marker. We convert the current last
+            // marker into a regular beat marker and use the current position
+            // as the new last marker.
+            markers.push_back(mixxx::BeatMarker(lastMarkerPosition,
+                    lastBeatLengthFrames(),
+                    lastMarkerBeatsPerBar));
+            lastMarkerPosition = position.toLowerFrameBoundary();
```

Comment:
> I have original expected that the change markers are quantized to beat markers.

Not currently, but I seem to recall we mentioned this somewhere that it could be a great follow up improvement (that is, to align to previous grid's beat if quantize=on)

Reply:
Captured [here](https://github.com/mixxxdj/mixxx/issues/13768#issuecomment-2440923542) for follow up work